source/blog/2015-09-22-rspec-compsable-matchers.md

```diff
@@ -0,0 +1,218 @@
+* `yield_successive_args`
+* `yield_with_args`
+* `yield_with_no_args`
+
+Let's take a look at what `values_match?` does.
+
+`values_match?` is the RSpec match definition that allows matchers, and other objects, to compose.
+The implementation looks like this:
+
+```ruby
+def self.values_match?(expected, actual)
+  if Hash === actual
+    return hashes_match?(expected, actual) if Hash === expected
+  elsif Array === expected && Enumerable === actual && !(Struct === actual)
+    return arrays_match?(expected, actual.to_a)
+  end
+
+  return true if expected == actual
+
+  begin
+    expected === actual
+  rescue ArgumentError
+    # Some objects, like 0-arg lambdas on 1.9+, raise
+    # ArgumentError for `expected === actual`.
+    false
+  end
+end
+```
+
+The first thing to note is the special casing for Hashes and Arrays. Basically,
+what those methods do is recursively apply `values_match?` to each of the
+elements in the various collection types as you'd expect. Arrays by scanning
+elements and hashes by matching keys and values.
+
+The next thing the method does is to return `true` if `expected == actual`,
+note not `expected === actual` but `==` because `/foo/ === /foo/` returns
+`false` as does `MyClass === MyClass` -- so we have to check normal equality
+as well and not rely solely on `===`. We can demonstrate this is back in our
+IRB session:
+
+```ruby
+>> expect(/foo/).to match(/foo/)
+=> true
+```
+
+What we've done here is constructed a both a target and match matcher with the value /foo/.
+When the match matcher hits `values_match?` it sees that /foo/ is not an array or a hash, and so
+checks `expected == actual`, which in this case is `/foo/ == /foo/`, which is true.
+
+The next line checks `expected === actual`. This is where the power of RSpec's new composability
+really kicks in. The reason for this is that the `BaseMatcher` class includes a module called
+`Composable`. The composable matcher defines `===` on matchers thusly:
+
+```rspec
+def ===(value)
+  matches?(value)
+end
+```
+
+This means that `expected` in the expression `expected === actual` can be
+another RSpec matcher, a regex, or any other object that responds to `===`.
+This means, for example, that one can trivially match a lambda by doing
+`expect(obj).to match(lambda { |actual| ... })`. Given that all built in RSpec
+matchers implement `===` we are able to compose them arbitrarily with other
+"matching" values in the Ruby system.
+
+An example is matching nesting hashes like so:
+
+```rspec
+expect({:data => {... complex hash ...}).to include(:data => a_hash_including(:response => "success"))
+```
+
+The reason we're able to pass the `a_hash_including` matcher inside the
+`include` matcher is that `values_match?` will call `===` on the inner
+`a_hash_including` matcher, which will invoke its `matches?` method, which
+will allow it to match the complex inner hash.
+
+##Conclusions
+
+The old RSpec match protocol used to be a lot easier to understand. Each of the
```

Comment:
I don't think the protocol was any easier to understand before.  The RSpec matcher protocol hasn't significantly changed.  It's just the internal implementation of many of our matchers has changed to delegate to this new fuzzy matching logic via `values_match?`.